### PR TITLE
[REVIEW] Add Series.Shift via Numba kernel

### PR DESCRIPTION
**Pull request by beckernick**
This PR adds the `shift` [functionality](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.shift.html) to numeric cudf.Series, analogous to pandas.

**Summary of Changes**
- Adds a `shift` method to cudf.Series, matches pandas API (with not yet implemented keyword arguments mandated to be their defaults)
- Adds a `gpu_shift` numba kernel in cudautils for out-of-place shifting
- Tests for forward and backward shifting of numeric dtypes
- Cell values that would become `NaN` based on the `period` in pandas shift are currently set to be `-1` in cudf shift.
- Requires that a column not contain any null values

This could serve as a stopgap until shift can be implemented in libcudf (#1285 ), as the performance is similar to standard binary operations on columns of millions of rows.

```
import cudf
import numpy as np

nelem = int(1e7)
df = cudf.DataFrame({'a':np.random.sample(nelem)})

%timeit df.a.shift(2)
%timeit df.a + 1
4.24 ms ± 132 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5.39 ms ± 137 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

### CODE REVIEW

**Comment by beckernick** [comment]
@kkraus14 , I'm seeing some interesting behavior with the numba kernel for negative `N` values and dataframe slices.

Given a dataframe `df`, if you slice the dataframe with `df.iloc[0:10]` and explicitly make a copy (whether internally to the method or external in the actual usage) before calling `series.shift(-1)`, everything works fine. If you don't explicitly make a copy, you end up overwriting the original series's indexes with `-4616189618054758400` as every value, which I take to be the placeholder value because the dtype is float64.

Curious if you have seen this before or have any insight. I'm wondering if it may be a copy vs view issue?

File: python/cudf/dataframe/series.py [hunk 1]
```diff
@@ -1479,6 +1481,28 @@ def digitize(self, bins, right=False):
...
+            raise NotImplementedError("Shift currently only supports "
+                                      "numeric dtypes")
+        if periods == 0:
+            return self
+
+        input_dary = self.data.to_gpu_array()
+        output_dary = rmm.device_array_like(input_dary)
+        cudautils.gpu_shift.forall(output_dary.size)(input_dary, output_dary,
+                                                     periods)
+        return Series(output_dary, name=self.name, index=self.index)
```
Comment (kkraus14) [inline python/cudf/dataframe/series.py]: This won't handle nulls in its current state. Suggest either adding a check for `null_count == 0` or implementing a bitshift to handle the nulls.

**Comment by kkraus14** [comment]
> Curious if you have seen this before or have any insight. I'm wondering if it may be a copy vs view issue? Making a copy of the buffer internally does resolve the issue and only adds a tiny bit of overhead (~4.2 ms becomes ~7 ms on average), but I'd like to better understand this behavior.

`iloc` will return a view as opposed to a copy by default. It should be fine to call the shift method against it as long as you're only setting data in the output array which should be a separate allocation.

File: python/cudf/utils/cudautils.py [hunk 2]
```diff
@@ -580,6 +580,24 @@ def gpu_unique_set_insert(vset, sz, val):
...
+def gpu_shift(in_col, out_col, N):
+    """Shift value at index i of an input array forward by N positions and
+    store the output in a new array.
+    """
+    i = cuda.grid(1)
+    if N > 0:
+        if i < in_col.size:
+            out_col[i] = in_col[i-N]
+        if i < N:
+            out_col[i] = -1
```
Comment (kkraus14) [inline python/cudf/utils/cudautils.py]: If we handle nulls, we can combine these two statements to just look like:
```
if i >=N and i < in_col.size:
```
that way we don't unnecessarily set the data where the nullmask bit would be set anyway.

**Comment by beckernick** [comment]
>>> iloc will return a view as opposed to a copy by default. It should be fine to call the shift method 
>>> against it as long as you're only setting data in the output array which should be a separate
>>> allocation.

This is what I thought, too. I don't appear to be setting any data in the input array, but can't seem to stop this behavior. See the following example. Setting `COPY_FLAG` will control whether to make a copy after using `iloc`.

COPY_FLAG as False:

```
import cudf
from cudf.dataframe import Series, DataFrame
import numpy as np
from numba import cuda
from librmm_cffi import librmm as rmm


COPY_FLAG = False

@cuda.jit
def gpu_shift(in_col, out_col, N):
    """Shift value at index i of an input array forward by N positions and
    store the output in a new array.
    """
    i = cuda.grid(1)
    if N > 0:
        if i < in_col.size:
            out_col[i] = in_col[i-N]
        if i < N:
            out_col[i] = -1
    else:
        if i <= (in_col.size + N):
            out_col[i] = in_col[i-N]
        if i >= (in_col.size + N):
            out_col[i] = -1

            
def shift(self, periods=1, freq=None, axis=0, fill_value=None):
    """Shift values of an input array by periods positions and store the
    output in a new array.
    Notes
    -----
    Shift currently only supports float and integer dtypes.
    """
    assert axis in (None, 0) and freq is None and fill_value is None

    if not np.issubdtype(self.dtype, np.number):
        raise NotImplementedError("Shift currently only supports "
                                  "numeric dtypes")
    if periods == 0:
        return self

    input_dary = self.data.to_gpu_array()
    output_dary = rmm.device_array_like(input_dary)
    gpu_shift.forall(output_dary.size)(input_dary, output_dary,
                                                 periods)
    return Series(output_dary, name=self.name, index=self.index)

np.random.seed(12)
nelem = int(1e4)
df = cudf.DataFrame({'a':np.random.sample(nelem)})

if COPY_FLAG:
    start = df.iloc[0:10].copy()
else:
    start = df.iloc[0:10]

print(start.head())
print(shift(start.a, -1).head())
print(start.a.head())
0     0.15416284237967237
1      0.7400496965154048
2     0.26331501518513467
3      0.5337393933802977
4    0.014574962485419674
Name: a, dtype: float64
-4616189618054758400      0.7400496965154048
-4616189618054758400     0.26331501518513467
-4616189618054758400      0.5337393933802977
-4616189618054758400    0.014574962485419674
-4616189618054758400       0.918747008099885
Name: a, dtype: float64
-4616189618054758400     0.15416284237967237
-4616189618054758400      0.7400496965154048
-4616189618054758400     0.26331501518513467
-4616189618054758400      0.5337393933802977
-4616189618054758400    0.014574962485419674
Name: a, dtype: float64
```

With COPY_FLAG as True:

```
import cudf
from cudf.dataframe import Series, DataFrame
import numpy as np
from numba import cuda
from librmm_cffi import librmm as rmm


COPY_FLAG = True

@cuda.jit
def gpu_shift(in_col, out_col, N):
    """Shift value at index i of an input array forward by N positions and
    store the output in a new array.
    """
    i = cuda.grid(1)
    if N > 0:
        if i < in_col.size:
            out_col[i] = in_col[i-N]
        if i < N:
            out_col[i] = -1
    else:
        if i <= (in_col.size + N):
            out_col[i] = in_col[i-N]
        if i >= (in_col.size + N):
            out_col[i] = -1

            
def shift(self, periods=1, freq=None, axis=0, fill_value=None):
    """Shift values of an input array by periods positions and store the
    output in a new array.
    Notes
    -----
    Shift currently only supports float and integer dtypes.
    """
    assert axis in (None, 0) and freq is None and fill_value is None

    if not np.issubdtype(self.dtype, np.number):
        raise NotImplementedError("Shift currently only supports "
                                  "numeric dtypes")
    if periods == 0:
        return self

    input_dary = self.data.to_gpu_array()
    output_dary = rmm.device_array_like(input_dary)
    gpu_shift.forall(output_dary.size)(input_dary, output_dary,
                                                 periods)
    return Series(output_dary, name=self.name, index=self.index)


np.random.seed(12)
nelem = int(1e4)
df = cudf.DataFrame({'a':np.random.sample(nelem)})

if COPY_FLAG:
    start = df.iloc[0:10].copy()
else:
    start = df.iloc[0:10]

print(start.a.head())
print(shift(start.a, -1).head())
print(start.a.head())
0     0.15416284237967237
1      0.7400496965154048
2     0.26331501518513467
3      0.5337393933802977
4    0.014574962485419674
Name: a, dtype: float64
0      0.7400496965154048
1     0.26331501518513467
2      0.5337393933802977
3    0.014574962485419674
4       0.918747008099885
Name: a, dtype: float64
0     0.15416284237967237
1      0.7400496965154048
2     0.26331501518513467
3      0.5337393933802977
4    0.014574962485419674
Name: a, dtype: float64
```

**Comment by beckernick** [comment]
Perhaps unrelated, but should the view from `iloc` have visibility into the full underlying data of the original object? I'm seeing that behavior. Passing the view into the numba kernel can result in data from outside the view going into the output array.

```
gdf = cudf.DataFrame({'a':[1,2,3,4,5,6,7]})
start = gdf.iloc[0:6]

input_dary = start.a.data.to_gpu_array()
output_dary = rmm.device_array_like(input_dary)
gpu_shift.forall(output_dary.size)(input_dary, output_dary,
                                             -1)
print(start)
print(Series(output_dary))
   a
0  1
1  2
2  3
3  4
4  5
5  6
0    2
1    3
2    4
3    5
4    6
5    7
dtype: int64
```

**Comment by kkraus14** [comment]
> Perhaps unrelated, but should the view from `iloc` have visibility into the full underlying data of the original object? I'm seeing that behavior. Passing the view into the numba kernel can result in data from outside the view going into the output array.
> 
> ```
> gdf = cudf.DataFrame({'a':[1,2,3,4,5,6,7]})
> start = gdf.iloc[0:6]
> 
> input_dary = start.a.data.to_gpu_array()
> output_dary = rmm.device_array_like(input_dary)
> gpu_shift.forall(output_dary.size)(input_dary, output_dary,
>                                              -1)
> print(start)
> print(Series(output_dary))
>    a
> 0  1
> 1  2
> 2  3
> 3  4
> 4  5
> 5  6
> 0    2
> 1    3
> 2    4
> 3    5
> 4    6
> 5    7
> dtype: int64
> ```

I was about to say the view likely still has access to the whole underlying device array allocation. Do the `size` and `shape` attributes look as expected? If so your numba kernel is reaching outside of where it should.

**Comment by beckernick** [comment]
Size and shape do look as expected, but it was an issue of grabbing too far. Resolved this issue by guarding in the kernel.

File: python/cudf/dataframe/series.py [hunk 3]
```diff
@@ -1479,6 +1481,33 @@ def digitize(self, bins, right=False):
...
+
+        Notes
+        -----
+        Shift currently only supports float and integer dtype columns with
+        no null values.
+        """
+        assert axis in (None, 0) and freq is None and fill_value is None
+
+        if self.null_count != 0:
+            raise AssertionError("Diff currently requires columns with no "
```
Comment (kkraus14) [inline python/cudf/dataframe/series.py]: Should this say "Shift currently requires columns with no null values"?

Comment (beckernick) [reply python/cudf/dataframe/series.py]: Yes, it should. Good catch. I should've caught that beforehand.

**Pull request merged**